### ISSUE TITLE
What's the earliest prepare() should reject?

### ISSUE THREAD

**Issue by jakearchibald** [WICG/shared-element-transitions#175]
```js
const readyPromise = transition.prepare(async () => {
  await domUpdate();
  await waitForFontsABit();
});
```

Let's say we know, before the callback is called, that the transition cannot happen. For example, due to duplicate `page-transition-tag`s.

We decided in https://github.com/WICG/shared-element-transitions/issues/169 that the callback should still be called, because the DOM change is more important than the transition, so the inability to create a transition shouldn't block the DOM change.

But, when should `readyPromise` reject? If we reject it as soon as possible, that may be before the callback is called. That seems wrong, but I'm really struggling to articulate why.

Some thoughts:

- Transitions are an enhancement to a DOM change. A failure in the transition shouldn't cancel/prevent the DOM update.
- The developer may wish to abandon work in the prepare callback when a newer transition supersedes it. However, this is better served by the Navigation API's [abort signal](https://wicg.github.io/navigation-api/#ref-for-dom-navigateevent-signal%E2%91%A0), or something more directly related to the DOM change rather than the transition.
- The transition API is interested in whether the DOM update fails. If it fails, it shouldn't attempt a transition.
- The developer is interested in whether the DOM update fails, as they may want to show an error.
- The developer is interested in when preparation completes successfully, but only to do transition-related things, like JS-based transition animations.
- The developer is interested in when the transition is complete, as they may want to delay particular things on the page until the transition is complete, or queue transitions.
- Errors in the `prepare` callback are likely to be application errors. Things like handling user input & data fetching should be handled before the `prepare` callback.
- If we choose to delay rejecting until 'after' the callback, what happens when the timeout is reached? Do we still reject while the callback is running in that case?
- If we discover that the transition cannot happen before running the callback, should we still block rendering during the transition callback? If not, then render-blocking is unreliable for the developer. Although, we can just provide a boolean to let them know if rendering is blocked.

Having written that down, I'm starting to convince myself that the `readyPromise` should reject as soon as possible, and we educate developers that they shouldn't write code that results in navigation/update failures should the transition fail.

**Comment by jakearchibald** [comment]
Copying over a comment by @khushalsagar from https://github.com/tabatkins/specs/pull/89#discussion_r937154280

---

Just to make sure, (the spec change in https://github.com/tabatkins/specs/pull/89) implies that a SameDocumentTransition is not reusable.  @vmpstr and I were discussing how to handle all the edge cases where there can be an error or a new transition while there is already an ongoing one and he made one argument for why its simpler to not allow reuse (its in the summary below).

Let's go through all the async points. What to do if we get another transition request at any of these points or there is an error.

1) prepare is called and we're waiting for the next rAF. This transition can't have an error but what to do if we get another prepare call (from a different transition object). The options are:

    - Add the previous callback to an ordered list. Assume that the transition is from current DOM to the state after executing all callbacks. So when capture finishes, all these callbacks will be executed serially to generate the incoming DOM.
    - Add the previous callback to an ordered list. Assume that the transition is from DOM after all pending callbacks. So execute all of them serially, rejecting promises and then start the capture phase. The latest callback runs to get the incoming DOM after capture is done.

1) At the next rAF we realize constraints are not satisfied so the transition needs to be aborted. I don't think we should suppress rendering in this case. So we can dispatch the callback, wait for it to finish and then reject promises. This would be equivalent to the developer synchronously updating the DOM without the feature.

2) We've issued an operation to capture snapshots and the operation fails. We immediately unblock rendering, tear down the pseudo DOM and dispatch the callback followed by resolving promises.

    If we get another prepare while we're waiting for the previous snapshot request to be serviced, then again we either assume transition is from the snapshotted state to the version after the final callback. So we add this prepare to the list that gets executed when the snapshot is done. Or abort the transition by unsuppressing rendering and executing all the pending callbacks. Then do another capture request and the transition continues as usual.

3) The snapshot operation is finished and the callback has been dispatched, we're waiting for it to resolve. If we get a prepares at this step, either assume transition is from currently snapshotted state to the version after the latest prepare. So add this to pending prepare callbacks and the transition starts once all the pending ones have finished executing.

    Or abort by unsupressing rendering and wait for the invoked callback to finish. The new PrepareCallback is the incoming DOM candidate. If we get another prepare in this state, add the current candidate to the list and update the incoming DOM candidate. Once all pending callbacks are done, capture begins and transition continues as usual.

4) The callback for the incoming DOM has resolved but the incoming DOM doesn't satisfy the constraints for the transition. Stop suppressing rendering and abandon the transition.

At any point, if a SameDocumentTransition object gets an abandon() call we can skip running the callback assuming that's what the dev wants. This way if the next prepare intentionally wants to abort the previous one, devs can call abandon on it. This is easier to reason about if we don't make the objects reusable.

**Comment by jakearchibald** [comment]
This is great analysis. I think we'll need a call to figure this one out.

I think we need to untangle which parts of this are about the transition, and which are about the DOM change / navigation.

I think we should assume that, if the DOM change is async, the developer will have systems to handle that sort of thing. Although, we shouldn't make it harder, and we should make it easy for developers to take transition state into account when handling navigations (eg, waiting for a current transition to end before attempting the next one).

For example, does abandoning the transition mean abandoning the navigation? I'm not sure it does. However, abandoning the navigation probably always means the transition should be abandoned.

We should also think about transitions that only impact part of the page. It seems like we have a similar issue if a parent element is the root of a transition when a child tries to become the root of a transition. I think the developer faces the same choice there, abort the parent transition vs waiting for it to complete.

Maybe we should offer an extra promise:

- `handled` - this is just the promise returned by the callback.
- `ready` - resolves when the transition is about to start, and the pseudo-elements are ready. Rejects when we know the transition cannot happen. It might reject before `handled` settles.
- `finished` - transition successfully completes.

In a Navigation API intercept, you'd await `handled`, since that feeds back the success of the navigation. Failure of the transition isn't really relevant to the navigation.

And maybe have something like `element.getClosestTransition()`, which returns the active transition on the current element, or the closest ancestor element, or the document, otherwise null. Developers would use this to optionally queue behind a running transaction, rather than force the closest transition to abandon.

**Comment by khushalsagar** [comment]
> For example, does abandoning the transition mean abandoning the navigation?

I see you're thinking through how a dev would integrate SET with the navigation API. That's good, I didn't have that perspective. It does make sense that abandoning a navigation also abandons the transition (if the user hits the stop button for example).

The other case I assumed a dev would use abandon is if updating the model state is disjoint from updating the view and the transition callback is limited to updating the view. So running each callback is redundant since each callback is updating the DOM based on current model state. This sounds a bit theoretical so please refute if necessary.

>  It seems like we have a similar issue if a parent element is the root of a transition when a child tries to become the root of a transition.

Ah. So this would be the case where the second prepare does want to use the current DOM as the outgoing one but it wants to change the tags so we need to capture again. You could do it by abandoning the current transition first right?

```js
let currentTransition;

async function triggerTransition() {
   // Maybe exposing the current phase value is good for this.
   // If the transition is already running let it finish first.
   if (currentTransition.running) {
     await currentTransition.finished;
   }

   // At this point either we're in the middle of the prepare callback
   // or it hasn't been dispatched yet.
   // We could either restore the DOM to the original outgoing state
   // and abandon the transition or let it finish.
   currentTransition.abandon();
   restoreDOMToView(currentView);

  currentTransition = SameDocumentTransition();
  // Start the new transition now.
}
```

I didn't follow at what point `handled` should resolve. But I think the idea of exposing the phase on SameDocumentTransition is trying to address the same gap?

**Comment by jakearchibald** [comment]
I'll flesh these notes out tomorrow:

- Figure out when the spinner would stop (before or after transition) - what would we recommend?
- With SET with different roots, these can be done by different components. Want option to queue or skip to end. Feels like queue should be in browser.
- If abandoned, should the prepare callback still be called? (talk about counter and header example)
- We need to make it easy to know when the DOM change happened (for the nav API)
- Keep navigation aborting and transition aborting separate
- Maybe transition abandon should be more like "skip to the end" (eg finish() on animations)

**Comment by jakearchibald** [comment]
## Transition vs navigation complete

Here's some best-practice code setting up a transition with the navigation API:

```js
navigation.addEventListener('navigate', (navigateEvent) => {
  const destinationURL = new URL(navigateEvent.destination.url);
  if (shouldNotIntercept(navigateEvent)) return;

  const signal = navigateEvent.signal;

  if (destinationURL.pathname.startsWith('/articles/')) {
    navigateEvent.intercept({
      async handler() {
        // Fetch content.
        const content = await getArticleContent(destinationURL.pathname, {
          signal,
        });

        // Perform transition
        const transition = new SameDocumentTransition();
        let navigationComplete;

        const performEndOfNavigation = async () => {
          await renderArticlePage(content, { signal });
          await waitForFontsABit({ signal });
        };

        await transition
          .prepare(async () => {
            navigationComplete = performEndOfNavigation();
            await navigationComplete;
          })
          .catch(() => {});

        await navigationComplete;
      },
    });
  }
});
```

The above is written assuming that the transition ready promise (returned by `prepare()`) may reject early, as soon as it knows the transition won't be possible. And also that the browser spinner should spin during the fetch related with the navigation, and stop once the DOM is changed (around the time the transition starts).

`navigateEvent.signal` will signal if the user presses the stop button, or a more important navigation begins.

The code is really tangled because the transition API needs to sit either side of the DOM change - it wants to know how long it takes, and if it fails. However, the navigation API also wants to know how long it takes, and if it fails, but it doesn't care if the transition fails.

One solution we've been exploring is delaying the resolution of the ready promise until the promise returned by the callback settles. But, this doesn't really help, as then you're left with a ready promise that rejects if the DOM change fails, or if the transition cannot happen. The navigation API doesn't care about the transition, so you're back to tangled code.

It also gets weird when you think about timeouts. As in, the transition preparation would timeout, but the promise returned by prepare would… ignore this timeout?

Instead, we should help the developer keep the transition and navigation as separate as possible. In this model, the ready promise and the `finished` promise refer to the transition only. They will reject early if the transition fails early. A new promise, `changeHandled`, resolves with the return value of the `prepare` callback, so:

```js
navigation.addEventListener('navigate', (navigateEvent) => {
  const destinationURL = new URL(navigateEvent.destination.url);
  if (shouldNotIntercept(navigateEvent)) return;

  const signal = navigateEvent.signal;

  if (destinationURL.pathname.startsWith('/articles/')) {
    navigateEvent.intercept({
      async handler() {
        // Fetch content.
        const content = await getArticleContent(destinationURL.pathname, {
          signal,
        });

        // Perform transition
        const transition = new SameDocumentTransition();

        transition.prepare(async () => {
          await renderArticlePage(content, { signal });
          await waitForFontsABit({ signal });
        });

        await transition.changeHandled;
      },
    });
  }
});
```

…now `changeHandled` gives the navigation API the bit it cares about. The ready promise and the `finished` promise don't need to be tangled up in this.

## Transition queues

My assumption is that, if/when we allow transitions to be scoped to something smaller than the whole document, two sibling elements that are transition roots should be able to overlap in terms of start & stop time. However, that won't be true if two transitions involve the same element in some way. So, if a parent element is involved in a transition, a new transition on a child element won't be able to happen at same time. And the vice versa is true, if a child element is involved in a transition, a new transition on a parent element won't be able to happen at the same time.

The new transition has two options, either queue behind the conflicting transitions, or clear out the conflicting transitions.

This is really hard to do in user-land, especially when dealing with two independent components, so we should probably have a feature for this. Eg:

```js
transition.prepare({
  skipPrevious: true,
  async handler() {
    // …
  }
});
```

I'm not sure what the default should be.

## Should the prepare callback ever be skipped?

I think we're in agreement that we should call `handler` even if we know in advance that the transition isn't possible due to something like duplicate `page-transition-tag`s. The transition is an enhancement to the DOM change, and a failure to do a transition shouldn't prevent the DOM change.

A trickier case is when a transition is abandoned in advance:

```js
const transition = new SameDocumentTransition();
transition.abandon();
transition.prepare(callback);
```

Also, a similar case, what if there are 5 transitions in the 'queue', and a new transition wants to `skipPrevious`?

I think we still want to run prepare callbacks for all the transitions in this case.

The example I'm thinking of: A counter component in a banner has a queued transition to switch from "1" to "2". Then, the banner wants to create a transition that moves the banner somewhere else on the page, and it wants to do so immediately. Just because the counter transition is skipped, it doesn't mean the counter shouldn't change from "1" to "2".

Abandoning DOM changes should be done through a separate mechanism, such as the navigation API's `signal`. So, even though all the prepare callbacks will be called, the developer may choose to immediately throw in those callbacks, to skip the DOM change.

Our naming of `transition.abandon()` might be misleading here. When a transition is abandoned, you're not left with the initial state, or some mid-state, you're left with the end state, so it's more of a "skip-to-the-end" than `abandon`. The animation API has `animation.finish()` which feels similar.

**Comment by khushalsagar** [comment]
A note from the spec review. If one of the constraints is not satisfied (like there are duplicate tags), it's unclear whether we should pause rendering for the duration of the PrepareCallback. We've already determined that this transition is invalid so seems odd to still pause rendering.

**Comment by jakearchibald** [comment]
Meeting notes:

- Have first promise that resolves with the callback
- Have another promise for ready
- Loose support for renaming abandon to finish
- Don't need first class support for signals
- Maybe don't block rendering if there are no `page-transition-tag`s, since it encourages using the feature for render blocking.
- No API for waiting for previous transition to fully animate
- Prefer to put options in creation mechanism
- Put the prepare handler into the constructor. Don't call it prepare.
- Constructor vs element method - pick one. Follow animation pattern. Some preference for factory function.

**Comment by khushalsagar** [comment]
This issue id tracking discussion for the spec [issue](https://tabatkins.github.io/specs/css-shared-element-transitions/#issue-a5efd1d7). We gotta drain the callbacks in the same order as the prepare callback.

And also [this](https://tabatkins.github.io/specs/css-shared-element-transitions/#issue-a8028177). We need the pending callbacks to run before capturing outgoing state for the new transition.

**Comment by jakearchibald** [comment]
Struggling with the naming on this one. Here's what I've got so far:

```js
const transition = document.createTransition({
  async updateDOM() {
    changeDOMSomehow();
  },
});
```

The argument is a dictionary so we can add other options later. It seems like WebIDL doesn't support unions where one type is a callback and the other is a dictionary, so there isn't an upgrade path here.

The assumption is we'll later have `element.createTransition()`.

- `transition.domUpdated`: Rejects if `updateDOM` rejects. Fulfills with `undefined` when `updateDOM` fulfills.
- `transition.ready`: Rejects if transition cannot start. Fulfills when pseduo-elements are created.
- `transition.finished`: Rejects if transition cannot start, or if it's skipped. Fulfills when the transition animation runs to completion.
- `transition.skipTransition()`: If the transition hasn't started, skip the animation (and reject `ready` and `finished` with an `AbortError`). If the transition is mid-way through, remove the pseuduo-elements and reject `finished` with an abort error. I get that `transition.skipTransition` is a bit tautologous, but I'm trying to suggest that the DOM change isn't skipped.

**Comment by jakearchibald** [comment]
I've got a spec PR for this https://github.com/tabatkins/specs/pull/97.

While writing it, I think we got something wrong in our plan:

> We need the pending callbacks to run before capturing outgoing state for the new transition.

I think this violates our "keep the transition and navigation separate" rule. Something other than our API should be in charge of queuing callbacks that should be queued.

The problem comes if a callback takes a long time to resolve. We would either:

- Queue indefinitely, meaning one bad callback can prevent all future queued DOM changes on the page.
- Timeout. But, in that case, the callback could still make DOM changes, so we're still allowing these callbacks to run in parallel.

Neither of these are good.

The navigation API doesn't perform any queuing, it relies on developers to manage that themselves. It feels like we should follow the same model.
